Framework/NexusGeometry/test/JSONGeometryParserTest.h

```diff
@@ -0,0 +1,424 @@
+    TS_ASSERT_EQUALS(translation, Eigen::Vector3d(0.971, 0, -0.049));
+    angleAxis = Eigen::AngleAxisd(parser.orientation(0));
+    TS_ASSERT_DELTA(angleAxis.angle(), parser.degreesToRadians(90), TOLERANCE);
+    TS_ASSERT_EQUALS(angleAxis.axis(), Eigen::Vector3d(0, 1, 0));
+    TS_ASSERT(!parser.isOffGeometry(0));
+    std::vector<Eigen::Vector3d> testVerticesVec{
+        {-0.001, 0, 0}, {0.001, 0.00405, 0}, {0.001, 0, 0}};
+    assertVectors(parser.vertices(0), testVerticesVec);
+    std::vector<uint32_t> testCylindersVec{0, 1, 2};
+    TS_ASSERT_EQUALS(parser.cylinders(0), testCylindersVec);
+    const auto &windingOrder = parser.windingOrder(0);
+    TS_ASSERT(windingOrder.empty());
+    const auto &faces = parser.faces(0);
+    TS_ASSERT(faces.empty());
+  }
+
+  void test_load_full_instrument_with_z_pixel_offset() {
+    std::string json =
+        Mantid::TestHelpers::getFullJSONInstrumentSimpleWithZPixelOffset();
+    JSONGeometryParser parser(json);
+    const auto &zPixelOffsets = parser.zPixelOffsets(0);
+    TS_ASSERT_EQUALS(zPixelOffsets.size(), 4);
+    TS_ASSERT_EQUALS(zPixelOffsets,
+                     (std::vector<double>{-0.0405, -0.0405, -0.0405, -0.0405}));
+  }
+
+private:
+  const double TOLERANCE = 1e-5;
+  std::string getSimpleJSONInstrument() { return ""; }
+
+  void attemptParseInvalidArgument(const std::string &json,
+                                   const std::string &expectedError) {
+    TS_ASSERT_THROWS_EQUALS((JSONGeometryParser(json)),
+                            const std::invalid_argument &e,
+                            std::string(e.what()), expectedError);
+  }
+
+  void assertVectors(const std::vector<Eigen::Vector3d> &lhs,
+                     const std::vector<Eigen::Vector3d> &rhs) {
+    TS_ASSERT(lhs.size() == rhs.size());
+    for (size_t i = 0; i < lhs.size(); ++i)
+      assertVertices(lhs[i], rhs[i]);
+  }
+
+  void assertVertices(const Eigen::Vector3d &lhs, const Eigen::Vector3d &rhs) {
+    TS_ASSERT_DELTA(lhs.x(), rhs.x(), TOLERANCE);
+    TS_ASSERT_DELTA(lhs.y(), rhs.y(), TOLERANCE);
+    TS_ASSERT_DELTA(lhs.z(), rhs.z(), TOLERANCE);
+  }
+};
+
+class JSONGeometryParserTestPerformance : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static JSONGeometryParserTestPerformance *createSuite() {
+    return new JSONGeometryParserTestPerformance();
+  }
+  static void destroySuite(JSONGeometryParserTestPerformance *suite) {
+    delete suite;
+  }
+
+  void setUp() override {
+    std::string filename = "V20_file_write_start_20190524.json";
+    std::string fullPath = ConfigService::Instance().getFullPath(
+        filename, true, Poco::Glob::GLOB_DEFAULT);
+
+    instrument = Strings::loadFile(fullPath);
+  }
+
+  void test_parse() {
+    TS_ASSERT_THROWS_NOTHING((JSONGeometryParser(instrument)));
+  }
+
+private:
+  std::string instrument;
+};
+
```

Comment:
Testing looks thorough 👍  for performance tests 👍  for so many destructive test cases